core/threshold/src/execution/endpoints/reshare_sk.rs

```diff
@@ -0,0 +1,1200 @@
+        },
+        tfhe_internals::{
+            compression_decompression_key::CompressionPrivateKeyShares,
+            glwe_key::GlweSecretKeyShare,
+            lwe_key::LweSecretKeyShare,
+            parameters::{DKGParams, DKGParamsBasics, DkgMode},
+            private_keysets::{
+                CompressionPrivateKeySharesEnum, GlweSecretKeyShareEnum, PrivateKeySet,
+            },
+        },
+    },
+};
+
+use tracing::instrument;
+
+pub struct ResharePreprocRequired {
+    pub batch_params_128: BatchParams,
+    pub batch_params_64: BatchParams,
+}
+
+impl ResharePreprocRequired {
+    /// Computes the number of randoms needed to reshare a private key set
+    /// where `num_parties_reshare_from` is the number of parties holding the input shares
+    /// (i.e. everyone in same set resharing, or the first set in two sets resharing)
+    pub fn new(num_parties_reshare_from: usize, parameters: DKGParams) -> Self {
+        let params = parameters.get_params_basics_handle();
+        let mut num_randoms_128 = 0;
+        let mut num_randoms_64 = 0;
+
+        num_randoms_64 += params.lwe_hat_dimension().0;
+
+        num_randoms_64 += params.lwe_dimension().0;
+
+        match parameters.get_params_basics_handle().get_dkg_mode() {
+            DkgMode::Z64 => {
+                num_randoms_64 += params.glwe_sk_num_bits() + params.compression_sk_num_bits()
+            }
+            DkgMode::Z128 => {
+                num_randoms_128 += params.glwe_sk_num_bits() + params.compression_sk_num_bits();
+                if let DKGParams::WithSnS(p) = parameters {
+                    num_randoms_128 += p.glwe_sk_num_bits_sns() + p.sns_compression_sk_num_bits();
+                }
+            }
+        }
+
+        num_randoms_128 *= num_parties_reshare_from;
+        num_randoms_64 *= num_parties_reshare_from;
+
+        ResharePreprocRequired {
+            batch_params_128: BatchParams {
+                triples: 0,
+                randoms: num_randoms_128,
+            },
+            batch_params_64: BatchParams {
+                triples: 0,
+                randoms: num_randoms_64,
+            },
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait ReshareSecretKeys {
+    async fn reshare_sk_same_set<
+        S: BaseSessionHandles,
+        P128: BasePreprocessing<ResiduePoly<Z128, EXTENSION_DEGREE>> + Send,
+        P64: BasePreprocessing<ResiduePoly<Z64, EXTENSION_DEGREE>> + Send,
+        const EXTENSION_DEGREE: usize,
+    >(
+        session: &mut S,
+        preproc128: &mut P128,
+        preproc64: &mut P64,
+        input_share: &mut Option<PrivateKeySet<EXTENSION_DEGREE>>,
+        parameters: DKGParams,
+    ) -> anyhow::Result<PrivateKeySet<EXTENSION_DEGREE>>
+    where
+        ResiduePoly<Z64, EXTENSION_DEGREE>: ErrorCorrect + Invert + Syndrome,
+        ResiduePoly<Z128, EXTENSION_DEGREE>: ErrorCorrect + Invert + Syndrome;
+
+    async fn reshare_sk_two_sets_as_s1<
```

Comment:
are these unit tested? I suppose it's not a big deal as we're testing the main logic in `reshare.rs` and we'll end up testing these when they're used in core/service.